README
Yaboot  -- PowerPC GNU/Linux OpenFirmware bootloader
-------------------------------------------------------------------

Please read the "COPYING" file for licence informations.

-------------------------------------------------------------------

Yaboot is an OpenFirmware bootloader for Open Firmware based
machines. It is known to work on "NewWorld" class powermacs (iMac and
all machines released after it), RS/6000, and possibly other OF based
CHRP machines. "OldWorld" PowerMacs (with the built-in MacOS ROM)
are not supported.

This README serves as a quick-start introduction to yaboot and its
installation utilities. For a more complete usage and installation
guide, see yaboot-howto.html/index.en.html in the same folder where
this README is located, or else inside the doc folder in this
directory. For documentation about all the options available, see the
man pages (bootstrap, yaboot, yaboot.conf, ybin, mkofboot, ofpath,
yabootconfig).


-------------------------------------------------------------------
Bootloader Installation Utilities  
ybin, mkofboot, ofpath, yabootconfig
Written by Ethan Benson <erbenson@alaska.net>
-------------------------------------------------------------------

ybin (YaBoot INstaller) and mkofboot
-------------------------------------------------------------------

ybin is a lilo/quik style bootloader installer for PowerPC based
machines which require a bootstrap partition rather than a traditional
bootblock (i.e. all `NewWorld' Macintoshes). When ybin is configured
correctly you can simply type ybin at the command line as root, and
the bootloader and its configuration file will be installed or updated
on the bootstrap partition without any further user intervention.

ybin also supports IBM PowerPC hardware which requires a slightly different
bootstrap partition setup. For these machines, ybin uses dd to write yaboot
directly to the partition instead of copying it to a filesystem on the
partition. ybin will add the requisite .note section to yaboot prior to
installation (IBM CHRP only).

mkofboot is a companion script (actually a symlink to ybin) which
initializes the bootstrap partition prior to running ybin to install
the bootloader on it. mkofboot will confirm you want to continue
before proceeding unless called with the -f or --force switch. This
is in contrast to ybin, which is non destructive except that it
overwrites any existing yaboot files (yaboot and yaboot.conf) at the
root level of the bootstrap filesystem.

mkofboot and ybin both refer to the same configuration file (normally
yaboot.conf) to determine where the bootstrap files will be placed. 

IMPORTANT: The bootstrap partition should never be mounted anywhere on
your filesystem, ybin and mkofboot will check if it is and refuse to
operate on it if it is mounted.  It is not necessary to keep anything
but the boot loader on the bootstrap partition, yaboot will load the
kernel from your root partition (which can be an ext2/3, XFS, or
ReiserFS filesystem). Do not mount the bootstrap partition on top of
/boot.

ybin can update a bootstrap filesystem either on a block device or in
an ordinary file (as in an image of a filesystem.) 

NOTE: You must have a secure mktemp program otherwise ybin will be
vulnerable to race conditions.  Debian's mktemp qualifies I don't know
about the other distributions, you have been warned. The temp file is
created in /tmp by default but ybin will respect the $TMPDIR
environment variable.


ofpath
-------------------------------------------------------------------

The included ofpath utility can usually determine the OpenFirmware
device path that corresponds with a unix device node in /dev/.  Ybin
uses this utility to find the path to the bootstrap partition and to
any defined macos/macosx partitions. ofpath is based on the utility
`show_of_path.sh' written by Olaf Hering.


NOTE: ofpath may not work with all SCSI cards/drivers.

IMPORTANT: ofpath will NOT work on NewWorld Powermacs if the machine
was booted with BootX.


yabootconfig
-------------------------------------------------------------------

Yabootconfig creates a default configuration file and then runs mkofboot to
complete the bootloader installation. yabootconfig reads the running
system's /etc/fstab to determine the kernel location, and detects the
location of the 800k Apple_Bootstrap partition. It will also find IBM
CHRP bootstrap partitions (type 0x41 PReP Boot).



Yaboot.conf Configuration File
-------------------------------------------------------------------

yaboot.conf configuration file settings are covered in detail in `man
yaboot.conf'. Avoid the use of spaces in the config file, except in
comment lines or inside quotes. Here are the essential settings.


boot=
Example:   boot=/dev/hda2

This required setting defines the bootstrap partition device.  It can
also be a regular file if you are creating a filesystem image for some
reason. The default config file has this set to ``unconfigured'' which
will cause ybin to complain about you not reading the docs, it is the
only option you should need to change for ybin to work. Be sure you
include the partition number at the end, it should never be something
like /dev/hda. 


install= 
Example:  install=/usr/lib/yaboot/yaboot or /usr/local/lib/yaboot/yaboot

The full pathname to the yaboot OpenFirmware executable file. The
default is shown above. This file will be copied to the root level of
the bootstrap partition. Note: If you are installing an alternative
bootloader, its filename will be changed to yaboot when it is copied
(to match what the first stage bootloader expects to be loading).


partition=
Example:   partition=3

The partition number on which the kernel image is located. By default,
this partition is assumed to be on the same device from which yaboot
was loaded. If needed, device= can be used to explicitly specify the 
device.


image=
Examples:  image=/vmlinux  or  image=/boot/vmlinux-2.2.19-pmac

The path to the image from the root level of the partition (remember
that yaboot is unaware of mountpoints).  Don't forget to include the
leading slash when specifying the image path (image=vmlinux will
probably fail).


root=
Example:   root=/dev/hda3

The Linux device name for the root partition. This parameter is
passed to the kernel when it starts up to let it know where its root
filesystem is located. Refer to the yaboot.conf man page for details
on kernel image options such as append=, initrd=, and initrd-size=.


magicboot= (same as -m or --magicboot)
Example:   /usr/local/lib/yaboot/ofboot 

Identifies the first stage loader. The example shown is the default. The
ofboot script it refers to is included with and configured automatically by
ybin from options in /etc/yaboot.conf. The magicboot= can be the full
pathname to any OF CHRP script file. Since some newer OpenFirmware
implementations appear to require a CHRP script, the magicboot= setting is
highly recommended. If you don't include magicboot=, then yaboot itself will
be the active bootfile.  magicboot= cannot be used on IBM CHRP.


Boot Menu Options
-------------------------------------------------------------------

The following options work only if you have specified the magicboot= option.
Each option specifies an OpenFirmware or unix device path to another
operating system's boot partition.  When you define one of these options you
will be presented with a simple menu at bootup allowing you to hit `l' to
boot GNU/Linux or another letter to boot the other OS (see below). These
letters are typed without a shift key (the boot menu is case sensitive).
When set to a unix device node such as /dev/hda11 then ybin will use the
ofpath utility to determine the OpenFirmware device path.

  Option       Letter  Operating System Type
  ---------    ------  ---------------------------------------------
  bsd=           b     NetBSD or OpenBSD root partition (you must 
                          have the BSD ofwboot bootloader 
                          installed at /usr/local/lib/yaboot/ofwboot)

  macos=         m     MacOS 8.* or 9.* bootable partition

  macosx=        x     MacOS X boot partition (also see brokenosx)

  darwin=        d     Darwin boot partition

  enablecdboot   c     Boot from the CDROM drive

  enablenetboot  n     Boot from the network

  enableofboot   o     Display an OpenFirmware prompt

  defaultos=           The default OS to load (linux, bsd, macos, 
                          darwin or macosx.

brokenosx

This option causes the menu entry for MacOSX to execute
\System\Library\CoreServices\BootX from the macosx=device instead of
the usual \\:tbxi.  This is necessary if OSX is installed onto an HFS+
filesystem instead of UFS. When OSX is installed on an HFS+ filesystem
MacOS will mount and debless the OSX partition.  Add this option if
the OSX menu entry breaks after booting MacOS.  You should not use
this option if OSX is installed on a UFS filesystem, for UFS installs
you specify the OSX bootstrap partition which is protected against
MacOS.

delay= 
Example:  delay=5

The time in seconds that the first stage ofboot loader will wait for
you to choose a letter before booting the default OS defined in
defaultos=.  If not set, the value of timeout= (converted to seconds)
will be used.


Colors
-------------------------------------------------------------------

fgcolor=string

Specifies the foreground (text) color used by yaboot and the
multiboot menu.  Available colors are: black, blue, light-blue, green,
light-green, cyan, light-cyan, red, light-red, purple, light- purple,
brown, light-gray, dark-gray, yellow, and white.  The default is
white.

bgcolor=string

Specifies the background color used by yaboot and the multiboot
menu.  Available colors are the same as fgcolor. The default is
black.


===========================================================================

Copyright (C) 2000, 2001 Ethan Benson

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

===========================================================================


ChangeLog
commit 58838f4a6973c60dc1b764fba20682ce07bb6c96
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Fri Aug 17 15:22:00 2007 +0100

    On some of the larger kernels we are starting to see a space squeeze.
    The kernel is currently being put at 20MBs and on some of the newer
    pSeries systems we are getting real-bases at 32MBs (plus AIX 5.3 has
    real-base at 32MB, so if AIX is ever booted, everything gets shifted).
    
    If the uncompressed kernel+initrd is larger than 12MBs then we see
    truncation (typical on install kernels).
    
    This patch moves the KERNELADDR to 0x00C00000 (12MB).  If real-base is
    12MB and is loaed there, yaboot will keep going up by a megabyte until
    prom_claim finds space.  And the uncompressed kernel should be
    indifferent where it's loaded because there's nothing at 32MB (ie it
    will operate as it does today).  If OF is loaded at 32MB then the 12MB
    region is free and the uncompressed kernel will have more space to grow.
    
    I have tested this on numerous pSeries boxes.  I have not tested on
    anything else.
    
    Signed-off-by: Jake Moilanen <moilanen@austin.ibm.com>

commit 1bea6140c2894b1259fd07f323b25fd846bd5acf
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu Jul 26 09:27:58 2007 +0100

    Fix typo courtesy of Kenichi Nagai

commit 58096e43b6efd2fb51e81866fff9660b70d5df64
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu Jul 19 14:16:30 2007 +0100

    Prepare 1.3.14

commit 6166753b1295119cfd8f7a5973b0b523098ef263
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu May 17 13:54:20 2007 +0100

    Netboot fixes:
    - the ARRAY_SIZE macro is no more needed here,
    - use of cfgpath to have the actual config path to apply for mac *and* ip
    lookup,
    - no need to free a buffer big enough, and besides malloc(9) was not
    enough ("/etc/" missed),
    - use of intermediate length variables to avoid unneeded calls to strlen()
    and strrchr().
    benoit.guillon <benoit.guillon@tele2.fr>

commit cfe3648f1f16d7399a2f0818ffb4e64da0c07cf5
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu May 17 11:30:59 2007 +0100

    Correct return type
    Leonardo Rangel <lrangel@linux.vnet.ibm.com>

commit 6084bb9a372a5fb9fa3e63a26c1770036c31883d
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu May 17 11:29:29 2007 +0100

    Add missing includes
    Leonardo Rangel <lrangel@linux.vnet.ibm.com>

commit 8aeef6f3fcbe661003268ee598de5601a192b965
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu May 17 11:28:19 2007 +0100

    Fix compilation warnings by using return not exit.
    Leonardo Rangel <lrangel@linux.vnet.ibm.com>

commit 1d2796d399a2076511df671e2f35eb0869085a6a
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu May 17 11:27:23 2007 +0100

    Now we stop truncating characters for pxelinux style boot at the directory seperator.
    Leonardo Rangel <lrangel@linux.vnet.ibm.com>

commit 06c6d687b3e76b2e728b1d59aaf408c20df029cb
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Wed May 16 10:22:59 2007 +0100

    Update ChangeLog from git

commit fa024941f8f1237e01e2ecf338442be9062953cc
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Tue May 15 14:28:02 2007 +0100

    use of cfg_set_default_by_mac() function in yaboot.c
    
    This last part calls the cfg_set_default_by_mac() function in yaboot.c
    by load_config_file() function.
    
    Signed-off-by: Paulo Ricardo Paz Vital <vital@br.ibm.com>

commit 313f71b25817a54832ac2ab46efcee2f5ab30331
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Tue May 15 14:25:55 2007 +0100

    This part creates the cfg_set_default_by_mac() function.
    
    This function sets the default cf_option if there is an image label if
    the same MAC address of the machine is netbooting.
    
    Signed-off-by: Paulo Ricardo Paz Vital <vital@br.ibm.com>

commit cf0768af118d79ba4888a7fb0fcd800fdf76dc98
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Tue May 15 14:24:06 2007 +0100

    load_my_config_file() modifications in yaboot.c
    
    This second part, deletes the netbooting like PXE algorithm from
    yaboot.c (reallocated in prom.c) and sets the use of prom_get_netinfo(),
    prom_get_mac (), prom_get_ip () functions in load_my_config_file()
    function to netbooting work like before the changes.
    
    Signed-off-by: Paulo Ricardo Paz Vital <vital@br.ibm.com>

commit 78cfabe0fac2682681c604bfe9b7266f173b9ea2
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Tue May 15 14:23:26 2007 +0100

    prom functions
    
    This first part creates the prom_get_netinfo (), prom_get_mac (),
    prom_get_ip () functions, and some structures to get the necessary
    information about the MAC and IP addresses.
    
    Actually, this is a reallocation of netbooting like PXE algorithm wrote
    by Benoit Guillon [1], to be used by any other part of Yaboot to get the
    netbooting information.
    
    [1] http://ozlabs.org/pipermail/yaboot-devel/2006-December/000062.html
    
    Signed-off-by: Paulo Ricardo Paz Vital <vital@br.ibm.com>

commit c5340d42bba9a51b5b0e22f2161985213d54f805
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Tue May 15 14:18:33 2007 +0100

    This patch fixes OF netbooting on pSeries
    Signed-off-by: Paulo Ricardo Paz Vital <vital@br.ibm.com>

commit 750ce1ac64bea60a4c072373ed9661c8b180b627
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Tue May 15 14:14:43 2007 +0100

    User-specified config file on Yaboot's prompt
    This patch clears out the kernel image list before loading a new config file.
    Leonardo Rangel lrangel at linux.vnet.ibm.com

commit 7305a1de1f811a7e5afe9ec24244848c298ef352
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu Apr 19 10:27:05 2007 +0100

    Enable user to supply config from yaboot command line.
    
    Leonardo Rangel <lrangel@linux.vnet.ibm.com>

commit aa67de3649c86c158622391aeec0016645b4f0d1
Author: Paul Nasrat <pnasrat@redhat.com>
Date:   Thu Apr 19 10:25:28 2007 +0100


second/cfg.c
/*
 *  cfg.c - Handling and parsing of yaboot.conf
 *
 *  Copyright (C) 1995 Werner Almesberger
 *                1996 Jakub Jelinek
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

#include "setjmp.h"
#include "stdarg.h"
#include "stdlib.h"
#include "string.h"
#include "types.h"
#include "prom.h"

/* Imported functions */
extern int strcasecmp(const char *s1, const char *s2);
extern int strncasecmp(const char *cs, const char *ct, size_t n);
extern char bootoncelabel[1024];

typedef enum {
     cft_strg, cft_flag, cft_end
} CONFIG_TYPE;

typedef struct {
     CONFIG_TYPE type;
     char *name;
     void *data;
} CONFIG;

#define MAX_TOKEN 200
#define MAX_VAR_NAME MAX_TOKEN
#define EOF -1

CONFIG cf_options[] =
{
     {cft_strg, "device", NULL},
     {cft_strg, "partition", NULL},
     {cft_strg, "default", NULL},
     {cft_strg, "timeout", NULL},
     {cft_strg, "password", NULL},
     {cft_flag, "restricted", NULL},
     {cft_strg, "message", NULL},
     {cft_strg, "root", NULL},
     {cft_strg, "ramdisk", NULL},
     {cft_flag, "read-only", NULL},
     {cft_flag, "read-write", NULL},
     {cft_strg, "append", NULL},
     {cft_strg, "initrd", NULL},
     {cft_flag, "initrd-prompt", NULL},
     {cft_strg, "initrd-size", NULL},
     {cft_flag, "pause-after", NULL},
     {cft_strg, "pause-message", NULL},
     {cft_strg, "init-code", NULL},
     {cft_strg, "init-message", NULL},
     {cft_strg, "fgcolor", NULL},
     {cft_strg, "bgcolor", NULL},
     {cft_strg, "ptypewarning", NULL},
     {cft_end, NULL, NULL}};

CONFIG cf_image[] =
{
     {cft_strg, "image", NULL},
     {cft_strg, "label", NULL},
     {cft_strg, "alias", NULL},
     {cft_flag, "single-key", NULL},
     {cft_flag, "restricted", NULL},
     {cft_strg, "device", NULL},
     {cft_strg, "partition", NULL},
     {cft_strg, "root", NULL},
     {cft_strg, "ramdisk", NULL},
     {cft_flag, "read-only", NULL},
     {cft_flag, "read-write", NULL},
     {cft_strg, "append", NULL},
     {cft_strg, "literal", NULL},
     {cft_strg, "initrd", NULL},
     {cft_flag, "initrd-prompt", NULL},
     {cft_strg, "initrd-size", NULL},
     {cft_flag, "pause-after", NULL},
     {cft_strg, "pause-message", NULL},
     {cft_flag, "novideo", NULL},
     {cft_strg, "sysmap", NULL},
     {cft_end, NULL, NULL}};

static char flag_set;
static char *last_token = NULL, *last_item = NULL, *last_value = NULL;
static int line_num;
static int back = 0;		/* can go back by one char */
static char *currp = NULL;
static char *endp = NULL;
static char *file_name = NULL;
static CONFIG *curr_table = cf_options;
static int ignore_entry;
static jmp_buf env;

static struct IMAGES {
     CONFIG table[sizeof (cf_image) / sizeof (cf_image[0])];
     int obsolete;
     struct IMAGES *next;
} *images = NULL;

void cfg_error (char *msg,...)
{
     va_list ap;

     va_start (ap, msg);
     prom_printf ("Config file error: ");
     prom_vprintf (msg, ap);
     va_end (ap);
     prom_printf (" near line %d in file %s\n", line_num, file_name);
     longjmp (env, 1);
}

void cfg_warn (char *msg,...)
{
     va_list ap;

     va_start (ap, msg);
     prom_printf ("Config file warning: ");
     prom_vprintf (msg, ap);
     va_end (ap);
     prom_printf (" near line %d in file %s\n", line_num, file_name);
}

inline int getc ()
{
     if (currp == endp)
	  return EOF;
     return *currp++;
}

#define next_raw next
static int next (void)
{
     int ch;

     if (!back)
	  return getc ();
     ch = back;
     back = 0;
     return ch;
}

static void again (int ch)
{
     back = ch;
}

static char *cfg_get_token (void)
{
     char buf[MAX_TOKEN + 1];
     char *here;
     int ch, escaped;

     if (last_token) {
	  here = last_token;
	  last_token = NULL;
	  return here;
     }
     while (1) {
	  while (ch = next (), ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r')
	       if (ch == '\n' || ch == '\r')
		    line_num++;
	  if (ch == EOF || ch == (int)NULL)
	       return NULL;
	  if (ch != '#')
	       break;
	  while (ch = next_raw (), (ch != '\n' && ch != '\r'))
	       if (ch == EOF)
		    return NULL;
	  line_num++;
     }
     if (ch == '=')
	  return strdup ("=");
     if (ch == '"') {
	  here = buf;
	  while (here - buf < MAX_TOKEN) {
	       if ((ch = next ()) == EOF)
		    cfg_error ("EOF in quoted string");
	       if (ch == '"') {
		    *here = 0;
		    return strdup (buf);
	       }
	       if (ch == '\\') {
		    ch = next ();
		    switch (ch) {
		    case '"':
		    case '\\':
			 break;
		    case '\n':
		    case '\r':
			 while ((ch = next ()), ch == ' ' || ch == '\t');
			 if (!ch)
			      continue;
			 again (ch);
			 ch = ' ';
			 break;
		    case 'n':
			 ch = '\n';
			 break;
		    default:
			 cfg_error ("Bad use of \\ in quoted string");
		    }
	       } else if ((ch == '\n') || (ch == '\r'))
		    cfg_error ("newline is not allowed in quoted strings");
	       *here++ = ch;
	  }
	  cfg_error ("Quoted string is too long");
	  return 0;		/* not reached */
     }
     here = buf;
     escaped = 0;
     while (here - buf < MAX_TOKEN) {
	  if (escaped) {
	       if (ch == EOF)
		    cfg_error ("\\ precedes EOF");
	       if (ch == '\n')
		    line_num++;
	       else
		    *here++ = ch == '\t' ? ' ' : ch;
	       escaped = 0;
	  } else {
	       if (ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r' || ch == '#' ||
		   ch == '=' || ch == EOF) {
		    again (ch);
		    *here = 0;
		    return strdup (buf);
	       }
	       if (!(escaped = (ch == '\\')))
		    *here++ = ch;
	  }
	  ch = next ();
     }
     cfg_error ("Token is too long");
     return 0;			/* not reached */
}

static void cfg_return_token (char *token)
{
     last_token = token;
}

static int cfg_next (char **item, char **value)
{
     char *this;

     if (last_item) {
	  *item = last_item;
	  *value = last_value;
	  last_item = NULL;
	  return 1;
     }
     *value = NULL;
     if (!(*item = cfg_get_token ()))
	  return 0;
     if (!strcmp (*item, "="))
	  cfg_error ("Syntax error");
     if (!(this = cfg_get_token ()))
	  return 1;
     if (strcmp (this, "=")) {
	  cfg_return_token (this);
	  return 1;
     }
     if (!(*value = cfg_get_token ()))
	  cfg_error ("Value expected at EOF");
     if (!strcmp (*value, "="))
	  cfg_error ("Syntax error after %s", *item);
     return 1;
}

static char *cfg_get_strg_i (CONFIG * table, char *item)
{
     CONFIG *walk;

     for (walk = table; walk->type != cft_end; walk++)
          if (walk->name && !strcasecmp (walk->name, item))
               return walk->data;
     return 0;
}

#if 0
// The one and only call to this procedure is commented out
// below, so we don't need this unless we decide to use it again.
static void cfg_return (char *item, char *value)
{
     last_item = item;
     last_value = value;
}
#endif

static int match_arch(const char *name)
{
	static prom_handle root;
	static char model[256], *p;

	if (!root) {
		if (!(root = prom_finddevice("/")))
			return 0;
	}

	if (!model[0]) {
		if (!prom_getprop(root, "compatible", model, sizeof(model)))
			return 0;
	}

	for (p = model; *p; p += strlen(p) + 1) {
		if (!strcasecmp(p, name))
			return 1;
	}

	return 0;
}

static void check_for_obsolete(const char *label)
{
	struct IMAGES *p;
	char *cur_label;

	/* Make sure our current entry isn't obsolete (ignored) */
	for (p = images; p; p = p->next) {
		if (curr_table == p->table && p->obsolete)
			return;
	}

	for (p = images; p; p = p->next) {
		if (curr_table == p->table)
			continue;

		cur_label = cfg_get_strg_i (p->table, "label");
		if (!cur_label)
			cur_label = cfg_get_strg_i (p->table, "image");

		if (!cur_label)
			continue;

		if (!strcasecmp(cur_label, label))
			p->obsolete = 1;
	}
}

static int cfg_set (char *item, char *value)
{
     CONFIG *walk;

     if (!strncasecmp (item, "image", 5)) {
	  struct IMAGES **p = &images;
	  int ignore = 0;

	  if (item[5] == '[' && item[strlen(item) - 1] == ']') {
		char *s, *q = item;

		/* Get rid of braces */
		item[strlen(item) - 1] = 0;
		item[5] = 0;

		for (s = item + 6; q; s = q) {
			q = strchr(s, '|');
			if (q)
				*q++ = 0;

			if (match_arch(s))
				goto cfg_set_cont;
                }
		/* This just creates an unused table. It will get ignored */
		ignore = 1;
	  } else if (item[5])
                goto cfg_set_redo;

cfg_set_cont:
	  while (*p)
	       p = &((*p)->next);
	  *p = (struct IMAGES *)malloc (sizeof (struct IMAGES));
	  if (*p == NULL) {
	       prom_printf("malloc error in cfg_set\n");
	       return -1;
	  }
	  (*p)->next = 0;
	  (*p)->obsolete = ignore;
	  curr_table = ((*p)->table);
	  memcpy (curr_table, cf_image, sizeof (cf_image));
     }

cfg_set_redo:
     for (walk = curr_table; walk->type != cft_end; walk++) {
	  if (walk->name && !strcasecmp (walk->name, item)) {
	       if (value && walk->type != cft_strg)
		    cfg_warn ("'%s' doesn't have a value", walk->name);
	       else if (!value && walk->type == cft_strg)
		    cfg_warn ("Value expected for '%s'", walk->name);
	       else {
		    if (!strcasecmp (item, "label"))
			 check_for_obsolete(value);
		    if (walk->data)
			 cfg_warn ("Duplicate entry '%s'", walk->name);
		    if (walk->type == cft_flag)
			 walk->data = &flag_set;
		    else if (walk->type == cft_strg)
			 walk->data = value;
	       }
	       break;
	  }
     }

     if (walk->type != cft_end)
	  return 1;

     //cfg_return (item, value);

     return 0;
}

static int cfg_reset ()
{
    CONFIG *walk;
#if DEBUG
    prom_printf("Resetting image table\n");
#endif
    line_num = 0;
    images = NULL;
    curr_table = NULL;
    curr_table = cf_options;
    for (walk = curr_table; walk->type != cft_end; walk++) {
#if DEBUG
        prom_printf("ItemA %s = %s\n", walk->name, walk->data);
#endif
        if (walk->data != NULL)
            walk->data = NULL;
#if DEBUG
        prom_printf("ItemB %s = %s\n\n", walk->name, walk->data);
#endif
    }

    return 0;
}

int cfg_parse (char *cfg_file, char *buff, int len)
{
     char *item, *value;

     file_name = cfg_file;
     currp = buff;
     endp = currp + len;

     cfg_reset();

     if (setjmp (env))
	  return -1;
     while (1) {
	  if (!cfg_next (&item, &value))
	       return 0;
	  if (!cfg_set (item, value)) {
#if DEBUG
	       prom_printf("Can't set item %s to value %s\n", item, value);
#endif
	  }
	  free (item);
     }
}

char *cfg_get_strg (char *image, char *item)
{
     struct IMAGES *p;
     char *label, *alias;
     char *ret;

     if (!image)
	  return cfg_get_strg_i (cf_options, item);
     for (p = images; p; p = p->next) {
	  if (p->obsolete)
		  continue;

	  label = cfg_get_strg_i (p->table, "label");
	  if (!label) {
	       label = cfg_get_strg_i (p->table, "image");
	       alias = strrchr (label, '/');
	       if (alias)
		    label = alias + 1;
	  }
	  alias = cfg_get_strg_i (p->table, "alias");
	  if (!strcmp (label, image) || (alias && !strcmp (alias, image))) {
	       ret = cfg_get_strg_i (p->table, item);
	       if (!ret)
		    ret = cfg_get_strg_i (cf_options, item);
	       return ret;
	  }
     }
     return 0;
}

int cfg_get_flag (char *image, char *item)
{
     return !!cfg_get_strg (image, item);
}

static int printl_count = 0;
static void printlabel (char *label, int defflag)
{
     int len = strlen (label);
     char a = ' ';

     if (!printl_count)
	  prom_printf ("\n");
     switch (defflag) {
	  case 1:  a='*'; break;
	  case 2:  a='&'; break;
	  default: a=' '; break;
     }
     prom_printf ("%c %s", a, label);
     while (len++ < 25)
	  prom_putchar (' ');
     printl_count++;
     if (printl_count == 3)
	  printl_count = 0;
}

void cfg_print_images (void)
{
     struct IMAGES *p;
     char *label, *alias;

     char *ret = cfg_get_strg_i (cf_options, "default");
     int defflag=0;

     printl_count = 0;
     for (p = images; p; p = p->next) {
	  if (p->obsolete)
		  continue;

	  label = cfg_get_strg_i (p->table, "label");
	  if (!label) {
	       label = cfg_get_strg_i (p->table, "image");
	       alias = strrchr (label, '/');
	       if (alias)
		    label = alias + 1;
	  }
	  if(!strcmp(bootoncelabel,label))
	       defflag=2;
	  else if(!strcmp(ret,label))
	       defflag=1;
	  else
	       defflag=0;
	  alias = cfg_get_strg_i (p->table, "alias");
	  printlabel (label, defflag);
	  if (alias)
	       printlabel (alias, 0);
     }
     prom_printf("\n");
}

char *cfg_get_default (void)
{
     char *label;
     struct IMAGES *p;
     char *ret = cfg_get_strg_i (cf_options, "default");

     if (ret)
	  return ret;
     if (!images)
	  return 0;

     for (p = images; p && p->obsolete; p = p->next);
     if (!p)
	     return 0;

     ret = cfg_get_strg_i (p->table, "label");
     if (!ret) {
	  ret = cfg_get_strg_i (p->table, "image");
	  label = strrchr (ret, '/');
	  if (label)
	       ret = label + 1;
     }
     return ret;
}

/*
 * cfg_set_default_by_mac ()
 * return 1 if the default cf_option was changed to label with the MAC addr
 * return 0 if not changed
 */
int cfg_set_default_by_mac (char *mac_addr)
{
     CONFIG *walk;
     struct IMAGES *tmp;
     char * label = NULL;
     int haslabel = 0;

     /* check if there is an image label equal to mac_addr */
     for (tmp = images; tmp; tmp = tmp->next) {
        label = cfg_get_strg_i (tmp->table, "label");
        if (!strcmp(label,mac_addr)){
            haslabel = 1;
        }
     }

     if (!haslabel)
         return 0;
     else {
         /*
          * if there is an image label equal to mac_addr, change the default
          * cf_options to this image label
          */
         for (walk = cf_options; walk->type != cft_end; walk++) {
             if (!strcasecmp(walk->name,"default")) {
                 walk->data = mac_addr;
                 return 1;
             }
         }
         return 0;
     }
}

/*
 * Local variables:
 * c-file-style: "k&r"
 * c-basic-offset: 5
 * End:
 */
